pypiserver/_app.py

```diff
@@ -289,14 +298,55 @@ def simple(project):
     if project != normalized:
         return redirect(f"/simple/{normalized}/", 301)
 
+    # pip install always uses 2 requests, first to get the list of versions
+    # and then to download the package.
+    # If we only search for the package on the local filesystem, we will
+    # might not be able to return the list containing the version.
+    # Hence, if pull_through is enabled, we will fetch the version list
+    # from pypi.org and then return the list of versions.
+
+    # If the packages version is not found, in pypi.org, we will return a 404
+
+    # Serving the actual package is handled by the /packages/ route
+    # And in this route, we will first attempt to fetch the package from the
+    # local filesystem, if not found, we will fetch it from pypi.org and save
+    # it to the local filesystem and then serve it.
+    if config.pull_through:
+        # Getting a list of packages and hashes
+        rgx = re.compile(
+            r"href=\"(?P<url>.*)#sha256=(?P<meta>.*\") ?>(?P<pkgversion>.*)</a><br />"
+        )
+        res = urllib.request.urlopen(
+            f"{config.fallback_url.rstrip('/')}/{project}/"
+        )
+        content = res.read().decode("utf-8").split("\n")
+        new_content = []
+        for line in content:
+            m = rgx.search(line)
+            if m:
+                gdict = m.groupdict()
+                vrsn = gdict["pkgversion"]
+                meta = gdict["meta"]
+                val = f'<a href="http://{config.host}:{config.port}/packages/{vrsn}#{meta}>{vrsn}</a><br />'
+            else:
+                val = line
+            new_content.append(val)
+        body = "\n".join(new_content)
+        response = HTTPResponse(body=body, status=200)
+        return response
+
     packages = sorted(
         config.backend.find_project_packages(project),
         key=lambda x: (x.parsed_version, x.relfn),
     )
+
     if not packages:
```

Comment:
[idea] here I'm also thinking, that to speed up the retrieval of `local U fallback` union of url sets (see comment above), we could run these requests in parallel with something like `asyncio.gather`. But I have to admit I honestly don't recall what is our state of supporting async code inside `pypi-server` right now, I'll need to refresh this 🙈😅

Reply:
As per my response above the process is pretty much sequential and fetching the packages using an async client isn't going to be faster.

We can do better by shortening the cycle by not having to asked pypi.org for metadata every time by caching it. But for that I'd like to do a new PR introducing a better storage backend...